### PR TITLE
Allow to password protect shares

### PR DESCRIPTION
**Pull request by alvaroaleman**
This changes allows to password protect shares. It works by:
* Allowing to optionally pass a password when creating a share
* If set, the password + salt that is configured via a new flag will be
  hashed via bcrypt and the hash stored together with the rest of the
  share
* Additionally, a random 96 byte long token gets generated and stored
  as part of the share
* When the backend retrieves an unauthenticated request for a share
   that has authentication configured, it will return a http 401
* The frontend detects this and will show a login prompt
* The actual download links are protected via an url arg that contains
  the previously generated token. This allows us to avoid buffering the
  download in the browser and allows pasting the link without breaking
  it

Thanks a lot for all the existing work on filebrowser btw, it's great!

Fixes #1223

### CODE REVIEW

File: http/public.go [hunk 1]
```diff
@@ -94,3 +104,26 @@ var publicDlHandler = withHashFile(func(w http.ResponseWriter, r *http.Request,
...
+	password := r.Header.Get("X-SHARE-PASSWORD")
+	if password == "" {
+		return http.StatusUnauthorized, errors.New("unauthorized")
+	}
+	if err := bcrypt.CompareHashAndPassword([]byte(l.PasswordHash), []byte(password+salt)); err != nil {
+		if errors.Is(err, bcrypt.ErrMismatchedHashAndPassword) {
+			return http.StatusUnauthorized, nil
+		}
+		return http.StatusBadRequest, err
+	}
```
Comment (ramiresviana) [inline http/public.go]: No errors are being returned.

File: frontend/src/views/Share.vue [hunk 2]
```diff
@@ -74,6 +74,23 @@
...
     <not-found v-if="error.message === '404'"></not-found>
     <forbidden v-else-if="error.message === '403'"></forbidden>
+    <div v-else-if="error.message === '401'">
+      <div class="card floating" id="password">
+        <div class="card-title">
+          <h2>{{ $t('buttons.password') }}</h2>
+        </div>
+
+        <div class="card-content">
+          <input v-focus type="password" placeholder="Password" v-model="password" @keyup.enter="fetchData">
```
Comment (ramiresviana) [inline frontend/src/views/Share.vue]: The `placeholder` attribute should use localized value.

File: frontend/src/components/prompts/Share.vue [hunk 3]
```diff
@@ -1,14 +1,11 @@
 <template>
-  <div class="card floating" id="share">
+  <div class="card floating" style="max-width: max-content; width: auto" id="share">
```
Comment (ramiresviana) [inline frontend/src/components/prompts/Share.vue]: Let's avoid inline styling.

File: frontend/src/components/prompts/Share.vue [hunk 4]
```diff
@@ -27,6 +24,13 @@
             :title="$t('buttons.copyToClipboard')"><i class="material-icons">content_paste</i></button>
         </li>
 
+        <li v-if="!hasPermanent">
+          <div style="text-align: right">
```
Comment (ramiresviana) [inline frontend/src/components/prompts/Share.vue]: Let's avoid inline styling.

**Review by ramiresviana — CHANGES_REQUESTED** [review]
This PR is not working correctly as the way it is, you need to do some testing. Also found some issues:

- Unprotected shares cannot be accessed.
- There's no feedback when user type password wrong.

**Comment by alvaroaleman** [comment]
> This PR is not working correctly as the way it is, you need to do some testing

@ramiresviana can you elaborate on that, what exactly is not working correctly? Or do you just mean the two issues you mentioned?

File: http/public.go (hunk 1 above)
Comment (alvaroaleman) [reply http/public.go]: Sorry, I don't understand the comment. Do you mean that I need to check for a non-zero statuscode above?

Comment (ramiresviana) [reply http/public.go]: When catching the `bcrypt.ErrMismatchedHashAndPassword` error, `nil` are being returned instead of `err`, allowing the request to be processed. That way protected shares can be accessed by providing a wrong password.

**Comment by ramiresviana** [comment]
> > This PR is not working correctly as the way it is, you need to do some testing
> 
> @ramiresviana can you elaborate on that, what exactly is not working correctly? Or do you just mean the two issues you mentioned?

The most important aspect of this feature was not working, the password validation. On my tests i was able to access any share by providing a wrong password. Also found another issue:

- Downloading selected files, using the select button on the header bar, is not possible. Looks like the download url is missing the token.

**Comment by alvaroaleman** [comment]
@ramiresviana addressed all feedback:
* Passwords are now properly checked
* Also added some tests for that
* Shares that are created without password now end up actually not having a password
* Entering a wrong password will result in an error being shown
* Selecting files and then downloading them via the button at the top works
* No inline styling
* Localized placeholder

Please have another look.

File: http/public.go (hunk 1 above)
Comment (alvaroaleman) [reply http/public.go]: An error here only indicates that there was an error doing the check, but you are right, I didn't check the returned status code of this func

File: frontend/src/views/Share.vue [hunk 5]
```diff
@@ -129,7 +148,7 @@ export default {
       return 'insert_drive_file'
     },
     link: function () {
-      return `${baseURL}/api/public/dl/${this.hash}${this.path}`
+      return `${baseURL}/api/public/dl/${this.hash}${this.path}?token=${this.token}`
```
Comment (ramiresviana) [inline frontend/src/views/Share.vue]: Let's omit the token parameter from the url on non-protected shares.

File: frontend/src/views/Share.vue [hunk 6]
```diff
@@ -193,8 +212,14 @@ export default {
       this.error = null
 
       try {
-        let file = await api.getHash(encodeURIComponent(this.$route.params.pathMatch))
+        const password = this.password || ''
+        if (password !== ''){
+          this.attemptedPasswordLogin = true
```
Comment (ramiresviana) [inline frontend/src/views/Share.vue:221]: `attemptedPasswordLogin` need to be added to the component state.

File: frontend/src/views/Share.vue [hunk 7]
```diff
@@ -193,8 +212,14 @@ export default {
       this.error = null
 
       try {
-        let file = await api.getHash(encodeURIComponent(this.$route.params.pathMatch))
+        const password = this.password || ''
```
Comment (ramiresviana) [inline frontend/src/views/Share.vue]: `password` need to be added to the component state.

File: frontend/src/views/Share.vue [hunk 8]
```diff
@@ -193,8 +212,14 @@ export default {
...
+        if (password !== ''){
+          this.attemptedPasswordLogin = true
+        }
+        let file = await api.getHash(encodeURIComponent(this.$route.params.pathMatch), password)
         this.path = file.path
+        this.token = file.token || ''
+        store.commit('setToken', this.token)
```
Comment (ramiresviana) [inline frontend/src/views/Share.vue]: That application store state is not being used, should be removed.

File: frontend/src/components/prompts/Share.vue [hunk 9]
```diff
@@ -130,7 +135,7 @@ export default {
     },
     getPermalink: async function () {
       try {
-        const res = await api.create(this.url)
+        const res = await api.create(this.url, this.passwordpermalink)
```
Comment (ramiresviana) [inline frontend/src/components/prompts/Share.vue]: `passwordpermalink` need to be added to the component state.

**Review by ramiresviana — CHANGES_REQUESTED** [review]
What about using JWT based authentication? I think would be a more secure approach.

File: frontend/src/views/Share.vue (hunk 8 above)
Comment (alvaroaleman) [reply frontend/src/views/Share.vue]: It is, it is used in the JS that generates the downloadlink when you select a file and click download in the top bar

File: frontend/src/components/prompts/Share.vue (hunk 9 above)
Comment (alvaroaleman) [reply frontend/src/components/prompts/Share.vue]: Can you give an example of what you mean by "component state"?

**Comment by alvaroaleman** [comment]
> What about using JWT based authentication? I think would be a more secure approach.

First, if you have conceptional remarks about a change, it would be awesome to give those in the first round of feedback, not in the second one. Second, can you elaborate on how exactly JWT would be more secure? The password is sent exactly once, that will be needed however auth works. The token can be any random value which must be hard to guess but making it a JWT or not will not make anything more or less secure.

**Comment by ramiresviana** [comment]
> > What about using JWT based authentication? I think would be a more secure approach.
> 
> First, if you have conceptional remarks about a change, it would be awesome to give those in the first round of feedback, not in the second one. Second, can you elaborate on how exactly JWT would be more secure? The password is sent exactly once, that will be needed however auth works. The token can be any random value which must be hard to guess but making it a JWT or not will not make anything more or less secure.

Just wanted to hear your opinion about JWT, that's not an requirement. I was worried about the fact that the share token is always the same, with the JWT standard the token needs to be refreshed periodically and the frontend ~do not have access to the payload~ (it has access to payload but cannot generate a valid token).

File: frontend/src/components/prompts/Share.vue (hunk 9 above)
Comment (ramiresviana) [reply frontend/src/components/prompts/Share.vue]: https://github.com/filebrowser/filebrowser/blob/f1b7bd59f67e719b7bfd203b0d7ec016fd21ab49/frontend/src/components/prompts/Share.vue#L69-L77

Some errors will appear on the console when `passwordpermalink` is not here.

File: frontend/src/views/Share.vue (hunk 8 above)
Comment (ramiresviana) [reply frontend/src/views/Share.vue]: I see, lets have it set on [/store/index.js](https://github.com/filebrowser/filebrowser/blob/master/frontend/src/store/index.js). Also use `this.$store.commit()` instead of importing the store.

**Comment by alvaroaleman** [comment]
> Just wanted to hear your opinion about JWT, that's not an requirement. I was worried about the fact that the share token is always the same, with the JWT standard the token needs to be refreshed periodically and the frontend do not have access to the payload (it has access to payload but cannot generate a valid token).

Well, I guess that depends on the expectation. Right now the download links from the share will work indefinitely for as long as that share exists, which is nice for example if you want to be able to pass around a link that will work without making the share public. If I wanted to brute force a filebrowser share, I would most likely try to brute force the password instead, because I would assume that it is almost always shorter than the 96 byte of the token and that it will in most cases be easier to guess.

File: frontend/src/components/prompts/Share.vue (hunk 9 above)
Comment (alvaroaleman) [reply frontend/src/components/prompts/Share.vue]: @ramiresviana fixed

File: frontend/src/views/Share.vue (hunk 8 above)
Comment (alvaroaleman) [reply frontend/src/views/Share.vue]: @ramiresviana fixed

File: frontend/src/views/Share.vue (hunk 7 above)
Comment (alvaroaleman) [reply frontend/src/views/Share.vue]: @ramiresviana fixed

File: frontend/src/views/Share.vue (hunk 6 above)
Comment (alvaroaleman) [reply frontend/src/views/Share.vue:221]: @ramiresviana fixed

File: frontend/src/views/Share.vue (hunk 5 above)
Comment (alvaroaleman) [reply frontend/src/views/Share.vue]: @ramiresviana fixed

**Review by ramiresviana — APPROVED** [review]

**Comment by alvaroaleman** [comment]
@o1egl Can you have a look, please?

File: .circleci/config.yml [hunk 10]
```diff
@@ -2,7 +2,7 @@ version: 2
 jobs:
   lint:
     docker:
-      - image: golangci/golangci-lint:v1.27.0
+      - image: golangci/golangci-lint:v1.31.0
```
Comment (o1egl) [inline .circleci/config.yml:5]: The last version is v1.36.0. Either upgrade to the latest or revert it.

File: cmd/root.go [hunk 11]
```diff
@@ -47,6 +47,7 @@ func init() {
 	flags.Bool("noauth", false, "use the noauth auther when using quick setup")
 	flags.String("username", "admin", "username for the first user when using quick config")
 	flags.String("password", "", "hashed password for the first user when using quick config (default \"admin\")")
+	flags.String("salt", "", "The salt to use when for hashing share password. Can be any value. If changed, existing password-protected shares wil stop working.") //nolint:lll
```
Comment (o1egl) [inline cmd/root.go]: Why this salt is needed? According to the https://en.wikipedia.org/wiki/Bcrypt#Description , the salt is automatically included in the output string which means there is no need to add it by yourself. 

```suggestion
	flags.String("salt", "", "The salt is used for hashing share passwords. It can be of any value. If changed, existing password-protected shares will stop working") //nolint:lll
```

File: frontend/src/api/files.js [hunk 12]
```diff
@@ -77,8 +77,13 @@ export function download (format, ...files) {
   if (format !== null) {
     url += `algo=${format}&`
   }
+  if (store.state.jwt !== ''){
+    url += `auth=${store.state.jwt}`
```
Comment (o1egl) [inline frontend/src/api/files.js]: If both jwt and token are not empty, the generated URL will be incorrect
```suggestion
    url += `auth=${store.state.jwt}&`
```

File: frontend/src/components/prompts/Share.vue [hunk 13]
```diff
@@ -72,7 +77,9 @@ export default {
       unit: 'hours',
       hasPermanent: false,
       links: [],
-      clip: null
+      clip: null,
+      password: '',
+      passwordPermalink: ''
```
Comment (o1egl) [inline frontend/src/components/prompts/Share.vue:82]: Why is it 2 different variables?

File: frontend/src/i18n/en.json [hunk 14]
```diff
@@ -142,7 +143,8 @@
     "show": "Show",
     "size": "Size",
     "upload": "Upload",
-    "uploadMessage": "Select an option to upload."
+    "uploadMessage": "Select an option to upload.",
+    "optionalPassword": "Optional Password"
```
Comment (o1egl) [inline frontend/src/i18n/en.json]:
```suggestion
    "optionalPassword": "Optional password"
```

File: frontend/src/views/Share.vue [hunk 15]
```diff
@@ -74,6 +74,24 @@
   <div v-else-if="error">
     <not-found v-if="error.message === '404'"></not-found>
     <forbidden v-else-if="error.message === '403'"></forbidden>
+    <div v-else-if="error.message === '401'">
+      <div class="card floating" id="password">
+        <div v-if="attemptedPasswordLogin" class="share__wrong__password {">{{ $t('login.wrongCredentials') }}</div>
```
Comment (o1egl) [inline frontend/src/views/Share.vue]:
```suggestion
        <div v-if="attemptedPasswordLogin" class="share__wrong__password">{{ $t('login.wrongCredentials') }}</div>
```

File: frontend/src/views/Share.vue [hunk 16]
```diff
@@ -203,6 +232,7 @@ export default {
         this.updateRequest(file)
         this.setLoading(false)
       } catch (e) {
+        console.log(e)
```
Comment (o1egl) [inline frontend/src/views/Share.vue]: Is it kept on purpose?

File: http/public.go [hunk 17]
```diff
@@ -94,3 +104,26 @@ var publicDlHandler = withHashFile(func(w http.ResponseWriter, r *http.Request,
 
 	return rawDirHandler(w, r, d, file)
 })
+
+func authenticateShareRequest(r *http.Request, l *share.Link, salt string) (int, error) {
```
Comment (o1egl) [inline http/public.go]: This is not the convenient way of writing go functions. Returning values on error is a bad practice

File: http/public.go [hunk 18]
```diff
@@ -94,3 +104,26 @@ var publicDlHandler = withHashFile(func(w http.ResponseWriter, r *http.Request,
...
+		return 0, nil
+	}
+
+	if r.URL.Query().Get("token") == l.Token {
+		return 0, nil
+	}
+
+	password := r.Header.Get("X-SHARE-PASSWORD")
+	if password == "" {
+		return http.StatusUnauthorized, nil
```
Comment (o1egl) [inline http/public.go:119]:
```suggestion
		return errors.New("password is required")
```

File: http/public.go [hunk 19]
```diff
@@ -19,6 +22,11 @@ var withHashFile = func(fn handleFunc) handleFunc {
 			return errToStatus(err), err
 		}
 
+		status, err := authenticateShareRequest(r, link, d.server.Salt)
+		if status != 0 || err != nil {
+			return status, err
+		}
```
Comment (o1egl) [inline http/public.go]:
```suggestion
		if err := authenticateShareRequest(r, link, d.server.Salt); err != nil {
			return http.StatusUnauthorized, err
		}
```

File: http/public_test.go [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+func TestPublicShareHandlerAuthentication(t *testing.T) {
+	t.Parallel()
+
+	const passwordBcrypt = "$2y$10$TFAmdCbyd/mEZDe5fUeZJu.MaJQXRTwdqb/IQV.eTn6dWrF58gCSe" //nolint:gosec
+	testCases := []struct {
+		name               string
+		share              *share.Link
+		req                *http.Request
+		expectedStatusCode int
```
Comment (o1egl) [inline http/public_test.go]: Please use `map[string]struct{}` for table tests definition. See https://github.com/filebrowser/filebrowser/blob/master/img/service_test.go#L20

File: http/public_test.go [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+				defer result.Body.Close()
+				if result.StatusCode != tc.expectedStatusCode {
+					t.Errorf("expected status code %d, got status code %d", tc.expectedStatusCode, result.StatusCode)
+				}
+			})
+		}
+	}
+}
+
+func newHTTPRequest(t *testing.T, m ...func(*http.Request)) *http.Request {
```
Comment (o1egl) [inline http/public_test.go]:
```suggestion
func newHTTPRequest(t *testing.T, requestModifiers ...func(*http.Request)) *http.Request {
```

File: http/share.go [hunk 22]
```diff
@@ -82,6 +88,10 @@ var sharePostHandler = withPermShare(func(w http.ResponseWriter, r *http.Request
 	rawExpire := r.URL.Query().Get("expires")
 	unit := r.URL.Query().Get("unit")
```
Comment (o1egl) [inline http/share.go]: Let's use JSON body to send all required params

File: http/share.go [hunk 23]
```diff
@@ -137,3 +164,34 @@ var sharePostHandler = withPermShare(func(w http.ResponseWriter, r *http.Request
 
 	return renderJSON(w, r, s)
 })
+
+func getSharePasswordHash(bodyReader io.Reader, salt string) (data []byte, statuscode int, err error) {
+	if bodyReader == nil {
+		return nil, 0, nil
+	}
+
+	bodyRaw, err := ioutil.ReadAll(bodyReader)
```
Comment (o1egl) [inline http/share.go]: Please use `json.NewDecoder()` instead

File: share/storage.go [hunk 24]
```diff
@@ -102,7 +102,9 @@ func (s *Storage) Gets(path string, id uint) ([]*Link, error) {
 			if err := s.Delete(link.Hash); err != nil {
 				return nil, err
 			}
-			links = append(links[:i], links[i+1:]...)
+			if len(links) > i+1 {
+				links = append(links[:i], links[i+1:]...)
+			}
```
Comment (o1egl) [inline share/storage.go]: Why this change is introduced?

File: users/storage.go [hunk 25]
```diff
@@ -17,6 +17,15 @@ type StorageBackend interface {
 	DeleteByUsername(string) error
 }
 
+type Interface interface {
```
Comment (o1egl) [inline users/storage.go]: Please give a proper name to this interface. `UserStore`?

**Review by o1egl — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 10 above)
Comment (alvaroaleman) [reply .circleci/config.yml:5]: It might not look like that, but the version was a very conscious choice, it had to be updated to something with go 1.15 to have support for `*testing.T.TempDir` which was added in go 1.15. This is the oldest version of the golangci-lint that uses go 1.15.

Updating it to v1.36.0 results in a bunch of new failures, because filebrowser doesn't have a golangci-lint config, so it uses whatever its default linters are, which is not stable across versions.

File: frontend/src/components/prompts/Share.vue (hunk 13 above)
Comment (alvaroaleman) [reply frontend/src/components/prompts/Share.vue:82]: Because its two different inputs

File: http/public.go (hunk 17 above)
Comment (alvaroaleman) [reply http/public.go]: An error here means something went wrong with our code and yes, in that case the status code can be ignored.  Statuscode != 0 means the verification failed. The distinction is important, because they result in different http codes. Our code erroring is a 5XX, the verification failing is a 4XX, if we return an error in both cases, we can not distinguish those two anymore

File: share/storage.go (hunk 24 above)
Comment (alvaroaleman) [reply share/storage.go]: Because a bunch of methods in storage.go unconditionally access `links[i+1:]` which results in an out of bounds panic if the last link expired. I will remove it from this PR.

File: users/storage.go (hunk 25 above)
Comment (alvaroaleman) [reply users/storage.go]: The linter will refuse that due to stuttering, package symbols should not start with the package name. This ends up being used as `users.Interface` which IMHO is totally reasonable

**Review by alvaroaleman — COMMENTED** [review]
@o1egl all feedback addressed, ptal

File: cmd/root.go (hunk 11 above)
Comment (alvaroaleman) [reply cmd/root.go]: Removed it

File: share/storage.go (hunk 24 above)
Comment (o1egl) [reply share/storage.go]: This check doesn't save from data race

File: users/storage.go (hunk 25 above)
Comment (o1egl) [reply users/storage.go]: I just gave an example. It can be named Store. An interface name should be meaningful and self explainable. `Interface` name is too generic.

File: share/storage.go (hunk 24 above)
Comment (alvaroaleman) [reply share/storage.go]: Sorry, what data race do you mean? It stops accessing an array in an unsafe manner and risking an out of bounds

File: users/storage.go (hunk 25 above)
Comment (alvaroaleman) [reply users/storage.go]: Renamed it to `Store`, please take another look

**Comment by alvaroaleman** [comment]
@o1egl any chance you can have another look?

**Comment by o1egl** [comment]
@alvaroaleman `Get Permanent Link` returns 400 error. It tries to decode an empty body

File: http/public.go (hunk 17 above)
Comment (o1egl) [reply http/public.go]: We already have a `errToStatus(err)` function. Let's add a mapping there

File: frontend/src/api/share.js [hunk 26]
```diff
@@ -23,14 +25,18 @@ export async function remove(hash) {
...
 
-export async function create(url, expires = '', unit = 'hours') {
+export async function create(url, password = '', expires = '', unit = 'hours') {
   url = removePrefix(url)
   url = `/api/share${url}`
   if (expires !== '') {
     url += `?expires=${expires}&unit=${unit}`
   }
-
+  let body = '';
```
Comment (o1egl) [inline frontend/src/api/share.js]: The server expects to receive a JSON. An empty string is not a correct JSON.

**Review by o1egl — CHANGES_REQUESTED** [review]

File: frontend/src/api/share.js (hunk 26 above)
Comment (alvaroaleman) [reply frontend/src/api/share.js]: @o1egl fixed

**Review by o1egl — APPROVED** [review]
